SmartPlaces.Facilities/lib/IngestionManager/src/AzureDigitalTwins/AdtGraphManager.cs

```diff
@@ -0,0 +1,382 @@
+        private readonly ParallelOptions parallelOptions;
+        private IngestionManagerOptions options;
+
+        public AdtGraphManager(ILogger<AdtGraphManager<TOptions>> logger,
+                               IOptions<TOptions> options,
+                               ITwinMappingIndexer twinMappingIndexer,
+                               TelemetryClient telemetryClient,
+                               bool skipUpload = false)
+        {
+            Logger = logger;
+            this.options = options.Value;
+            TelemetryClient = telemetryClient;
+            TwinMappingIndexer = twinMappingIndexer;
+            SkipUpload = skipUpload;
+
+            for (int i = 0; i < this.options.MaxDegreeOfParallelism; i++)
+            {
+                queue.Enqueue(new DigitalTwinsClient(new Uri(options.Value.AzureDigitalTwinsEndpoint), new DefaultAzureCredential()));
+            }
+
+            parallelOptions = new ()
+            {
+                MaxDegreeOfParallelism = options.Value.MaxDegreeOfParallelism,
+            };
+        }
+
+        protected ILogger Logger { get; }
+
+        protected TelemetryClient TelemetryClient { get; }
+
+        protected ITwinMappingIndexer TwinMappingIndexer { get; }
+
+        protected bool SkipUpload { get; }
+
+        public async Task<IEnumerable<string>> GetModelAsync(CancellationToken cancellationToken)
+        {
+            // Get all of the Models to be used for creating instances of twins
+            if (queue.TryDequeue(out var digitalTwinsClient))
+            {
+                var modelsResults = digitalTwinsClient.GetModelsAsync(new GetModelsOptions { IncludeModelDefinition = true }, cancellationToken);
+                var modelList = new List<string>();
+                queue.Enqueue(digitalTwinsClient);
+
+                await foreach (DigitalTwinsModelData md in modelsResults)
+                {
+                    // TODO: jobee - Add error handling for models which do not have a DtdlModel (is this even possible?)
+                    if (md.DtdlModel != null)
+                    {
+                        modelList.Add(md.DtdlModel);
+                    }
+                }
+
+                return modelList;
+            }
+            else
+            {
+                throw new Exception("Unable to get ADT Client instance");
+            }
+        }
+
+        public virtual async Task UploadGraphAsync(Dictionary<string, BasicDigitalTwin> twins, Dictionary<string, BasicRelationship> relationships, CancellationToken cancellationToken)
+        {
+            if (!SkipUpload)
+            {
+                await ImportTwinsAsync(twins);
+                await ImportRelationshipsAsync(relationships);
+            }
+            else
+            {
+                Logger.LogInformation("Skip upload is set to true. Skipped loading ADT Instance.");
+            }
+
+            await UpdateCache(twins);
```

Comment:
Not everyone wants the cache. We should probably add an options flag to opt-in to using the cache